Adding create currencies command. Updating base Currency class

remedge opened the pull request:
- add command to create base currencies;
- add methods to base Currency class;
- add ProviderChain;
- using own ProviderFactory to use ProviderChain;
- using own Converter with own ProviderFactory;
- update README.

remedge committed:
update bundle for work

- add command to create base currencies;
- add methods to base Currency class;
- add ProviderChain;
- using own ProviderFactory to use ProviderChain;
- using own Converter with own ProviderFactory;
- update README.

maZahaca commented:
Thanks for PR!

remedge committed:
update CreateBaseCurrenciesCommand

transfer main CreateBaseCurrencies-functionality to CurrencyManager.

remedge committed:
remove extra CurrencyConverter

remedge commented:
> Please add functionality for adding Currencies through the CurrencyManager

done.

> Why that class has been copied here? Please use CurrencyConverter instead

Sorry, my bad, thought that's the same story as with ProviderFactory. Fixed.

> What about old compiler RegisterProvidersPass? I don't see any deletion of it. Looks like that class does the same thing

Old compiler didn't work, because ProviderFactory don't have method addProvider(). It seems like there was an attempt to the use classical symfony tagged services case, but some necessary code is missed. So i rewrote this case in more canonical form, as [docs describes](http://symfony.com/doc/current/components/dependency_injection/tags.html).  And its quite possible to remove old compiler, because its unused now.

> Duplicate! Use ProviderFactory

As i said, old ProviderFactory did't load the providers because of some missed code (you can check it if you try to load currency rates with console command). In efforts to stay with PR only in one repo, i guess its better idea to use own ProviderFactory.

> specify param types

Which ones? Now it simply loads all currencies, that are used in cbr and ecb providers.

> What is the purpose of adding ProviderChain? I don't see any point of it. Could you clarify.

As i also said, its just a canonical example from documentation. It think its nice and useful. But for decomposing the base `redco/redcode-currency-rate` from `redco/redcode-currency-rate-bundle` it might not so good idea. As you wish.

Btw, i think its a good way - to not use in base repo `redco/redcode-currency-rate` parts from `redco/redcode-currency-rate-bundle`. Otherwise, why there is this decomposition at all? I want to say, i think that it's possible to improve `redco/redcode-currency-rate` also :)

maZahaca commented:
> Old compiler didn't work, because ProviderFactory don't have method addProvider(). It seems like there was an attempt to the use classical symfony tagged services case, but some necessary code is missed. So i rewrote this case in more canonical form, as docs describes. And its quite possible to remove old compiler, because its unused now.

I think that had happened because miss-versioning. 
There're no code with addProvider() in the master, but it's existed in the [0.2](https://github.com/redco/redcode-currency-rate/blob/0.2/src/RedCode/Currency/Rate/Provider/ProviderFactory.php) branch.
I merged it to the master. Current component has dependency to stable "redcode/currency-rate": "0.2.*". Just try to update you component.

> As i said, old ProviderFactory did't load the providers because of some missed code (you can check it if you try to load currency rates with console command). In efforts to stay with PR only in one repo, i guess its better idea to use own ProviderFactory

I've just tried on the latest version. I missed creating of new tag for latest code on currency lib. Now this is fixed!
Try to update to the last version with request:

``` bash
composer update redcode/currency-rate-bundle redcode/currency-rate
```

> its better idea to use own ProviderFactory.
> If you want to change it somehow, you can create separated PR to the [redcode-currency-rate](https://github.com/redco/redcode-currency-rate)
> But the _BEST_ practice is to announce your upcoming change or issue, to discuss it first.
> 
> As i also said, its just a canonical example from documentation

I don't see any profit of using ProviderChain, in that particular lib it just covers array access, and do the same as was in the ProviderFactory before.
So, i don't think this is a good idea to split that code, if don't have another concern.

> Btw, i think its a good way - to not use in base repo redco/redcode-currency-rate parts from redco/redcode-currency-rate-bundle

The main reason to use such way was to determine business logic from any adoption to any framework.
I used redco/redcode-currency-rate in Symfony projects but also in others which don't have any framework.

After all of this I will:
- pay more attention on releasing new versions to avoid such errors.
- add code style requirements to [Symfony](http://symfony.com/doc/current/contributing/code/standards.html) on this project
- add CONTRIBUTION notes
- it would be great to add some tests to avoid troubles with stability

remedge committed:
rollback to stable version changes

- remove duplicate ProviderCompiler;
- remove ProviderChain;
- remove own ProviderFactory;
- remove own ProviderException;
- fix typo in README;

remedge committed:
rollback changes for services.xml

remedge committed:
god damn rollback!

remedge committed:
jesus christ rollback!

remedge commented:
> I merged it to the master. Current component has dependency to stable "redcode/currency-rate": "0.2.*". Just try to update you component.

I updated the packages, and `redcode:currency:rate:load` works fine now, thank you. 

In this case, there is no need to my ProviderFactory updates. And only profit of this PR is:
- adding of creating currencies command;
- add getters and setters to Entity/Currency class (README examples didn't works without that)

> If you want to change it somehow, you can create separated PR to the redcode-currency-rate
> But the BEST practice is to announce your upcoming change or issue, to discuss it first.

Thanks, I'll take this practice for the future :)

maZahaca commented:
Ok, clean all other code except this:
- adding of creating currencies command;
- add getters and setters to Entity/Currency class (README examples didn't works without that)

remedge commented:
it's already done

maZahaca commented:
Don't see any changes. You last commit was authored 3 days ago

remedge commented:
Last commits were Sep 29, 2015 0_o
Check [changed files](https://github.com/redco/redcode-currency-rate-bundle/pull/6/files) - there must be only 4 files

maZahaca commented:
So you need to update this PR somehow, because it's not in it.

> update bundle for work …
> remedge authored 3 days ago

maZahaca commented:
I've added code style requirements to [Symfony Coding Standards](http://symfony.com/doc/current/contributing/code/standards.html). Check out last version please

remedge committed:
Merge remote-tracking branch 'redco/master'

Conflicts:
	Entity/Currency.php

maZahaca commented on Entity/Currency.php:
```diff
@@ -9,4 +9,13 @@
 {
+    public function setCode($code)
+    {
+        $this->code = $code;
```
must contains return $this;

To simplify such decorations: 

``` php
$currency->setCode()->setSomethingElse()
```

maZahaca commented on .gitignore:
```diff
@@ -1 +0,0 @@
-vendor
```
Why did you delete this?

maZahaca commented:
Now before commit you can check you code style and test coverage by running the command:

```
./Tests/runTests.sh
```

If it shows some issues, please fix them first.
I will add this requirement to the Contribution notes

maZahaca commented:
Also it's configured to post PR hook for checking all this things.
So you can see if it's okay after some time in the PR

maZahaca commented:
I've added [CONTRIBUTION](https://github.com/redco/redcode-currency-rate-bundle/blob/master/CONTRIBUTING.md) guide. 
Sorry for lots of changes, but this is necessary to increate the quality of the code! :cactus:

remedge committed:
Merge remote-tracking branch 'redco/master'

remedge committed:
restore .gitignore; fix code style.

remedge commented:
I highly appreciate your high code standards :+1: 
It's all fixed.

maZahaca commented on Manager/CurrencyManager.php:
```diff
@@ -40,4 +49,29 @@ public function getAll()
+     * @throws ORMInvalidArgumentException
+     * @throws OptimisticLockException
+     */
+    public function createBaseCurrencies()
```
> Please add functionality for adding Currencies through the CurrencyManager

I actually meant following: 
- add CurrencyManager::addCurrency($code) which will cover EntityManager calls
- add ICurrencyManager::addCurrency($code)

And most of this code will stay in the Command, it's not necessary for CurrencyManager to know about some base currencies

remedge-4xxi committed:
rollback save functionality to command.

remedge commented:
Check my commit please - did I understand you correctly? If yes, i will make PR to redcode-currency-rate with ICurrencyManager changes.

remedge commented:
I have some difficulties with testing addCode() method cause of `new $this->currencyClassName()` construction. To resolve this, i suggest to rewrite some code parts - [in this issue](https://github.com/redco/redcode-currency-rate-bundle/issues/13).

maZahaca commented:
addCode -> addCurrency($code)

remedge committed:
fix addCode -> addCurrency

remedge committed:
optimize createCurrenciesCommand
